### ISSUE TITLE
DefaultQuantParamsPass doesn't work correctly if bias constant has multiple users

### ISSUE THREAD

**Issue by lgeiger** [tensorflow/tensorflow#40055]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.3.0.dev20200601 
- Python version: 3.7.7
- Bazel version (if compiling from source): 3.0.0

**Describe the current behavior**

[`DefaultQuantParamsPass`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/transforms/default_quant_params.cc) doesn't correctly handle quantization if a  `bias` constant has multiple users. This issue can be seen in the following toy model:
```python
img = tf.keras.layers.Input((32, 32, 3))
x = tf.keras.layers.Conv2D(64, 3, activation="relu", use_bias=True)(img)
x = tf.keras.layers.BatchNormalization()(x)
model = tf.keras.models.Model(img, x)
```
Since the `bias` of the `Conv2D` and `beta` from the `BatchNormalization` are both zero at initialization and have the same shape these tensors are shared by MLIR and don't convert correctly since `DefaultQuantParamsPass` doesn't duplicate the tensor to satisfy the differet quantization constraints by `tfl.conv_2d` and `tfl.mul`.

**Describe the expected behavior**
The `bias` of the `Conv2D` layer should be quantized to `int32` as required by `tfl.conv_2d`  and `beta` used by `tfl.mul` from the `BatchNormalization` layer should be quantized to `int8` as required my `tfl.mul`.

**Standalone code to reproduce the issue**
The issue can be reproduced by the following MLIR test:
```mlir
// RUN: tf-opt %s --tfl-default-quant --tfl-quantize --tfl-post-quantize | FileCheck %s  --dump-input-on-failure

// CHECK-LABEL: test_quantize_shared_bias
func @test_quantize_shared_bias(%arg0: tensor<1x112x112x32xf32>, %arg1: tensor<32x3x3x3xf32>) -> tensor<1x56x56x32xf32> {
  %cst = constant dense<0.0> : tensor<32xf32>
  %conv = "tfl.conv_2d"(%arg0, %arg1, %cst) {dilation_h_factor = 2 : i32, dilation_w_factor = 3 : i32, fused_activation_function = "RELU", padding = "SAME", stride_h = 4 : i32, stride_w = 5 : i32} : (tensor<1x112x112x32xf32>, tensor<32x3x3x3xf32>, tensor<32xf32>) -> tensor<1x56x56x32xf32>
  %add = "tfl.add"(%conv, %cst) {fused_activation_function = "NONE"} : (tensor<1x56x56x32xf32>, tensor<32xf32>) -> tensor<1x56x56x32xf32>
  return %add : tensor<1x56x56x32xf32>

  // CHECK: %0 = "tfl.pseudo_qconst"() {qtype = tensor<32x!quant.uniform<i32:f32, 6.1514801999231058E-5>>, value = dense<0> : tensor<32xi32>} : () -> tensor<32x!quant.uniform<i32:f32, 6.1514801999231058E-5>>
  // CHECK: %1 = "tfl.pseudo_qconst"() {qtype = tensor<32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>, value = dense<-128> : tensor<32xi8>} : () -> tensor<32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>
  // CHECK: %2 = "tfl.conv_2d"(%arg0, %arg1, %0) {dilation_h_factor = 2 : i32, dilation_w_factor = 3 : i32, fused_activation_function = "RELU", padding = "SAME", stride_h = 4 : i32, stride_w = 5 : i32} : (tensor<1x112x112x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>, tensor<32x3x3x3x!quant.uniform<u8:f32, 0.0078431372549019607:128>>, tensor<32x!quant.uniform<i32:f32, 6.1514801999231058E-5>>) -> tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>
  // CHECK: %3 = "tfl.add"(%2, %1) {fused_activation_function = "NONE"} : (tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>, tensor<32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>) -> tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>
  // CHECK: return %3 : tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>
}
```

The current implementation outputs the following graph which fails to quantize the convolution:
```mlir
func @test_quantize_shared_bias(%arg0: tensor<1x112x112x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>, %arg1: tensor<32x3x3x3x!quant.uniform<u8:f32, 0.0078431372549019607:128>>) -> tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>> {
  %cst = constant dense<0.000000e+00> : tensor<32xf32>
  %0 = "tfl.dequantize"(%arg1) : (tensor<32x3x3x3x!quant.uniform<u8:f32, 0.0078431372549019607:128>>) -> tensor<32x3x3x3xf32>
  %1 = "tfl.dequantize"(%arg0) : (tensor<1x112x112x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>) -> tensor<1x112x112x32xf32>
  %2 = "tfl.conv_2d"(%1, %0, %cst) {dilation_h_factor = 2 : i32, dilation_w_factor = 3 : i32, fused_activation_function = "RELU", padding = "SAME", stride_h = 4 : i32, stride_w = 5 : i32} : (tensor<1x112x112x32xf32>, tensor<32x3x3x3xf32>, tensor<32xf32>) -> tensor<1x56x56x32xf32>
  %3 = "tfl.quantize"(%2) {qtype = tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>} : (tensor<1x56x56x32xf32>) -> tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>
  %4 = "tfl.dequantize"(%3) : (tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>) -> tensor<1x56x56x32xf32>
  %5 = "tfl.add"(%4, %cst) {fused_activation_function = "NONE"} : (tensor<1x56x56x32xf32>, tensor<32xf32>) -> tensor<1x56x56x32xf32>
  %6 = "tfl.quantize"(%5) {qtype = tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>} : (tensor<1x56x56x32xf32>) -> tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>
  return %6 : tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>
}
```

When allowing hybrid operands the issue becomes clearer and it generates a graph that will fail to execute in TFLite due to a floating point bias being used in the quantized convolution:
https://github.com/tensorflow/tensorflow/blob/de32c75f2f0b9c298d858180fc19fa8881bfab41/tensorflow/compiler/mlir/lite/transforms/quantize.cc#L72
```mlir
func @test_quantize_shared_bias(%arg0: tensor<1x112x112x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>, %arg1: tensor<32x3x3x3x!quant.uniform<u8:f32, 0.0078431372549019607:128>>) -> tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>> {
  %cst = constant dense<0.000000e+00> : tensor<32xf32>
  %0 = "tfl.conv_2d"(%arg0, %arg1, %cst) {dilation_h_factor = 2 : i32, dilation_w_factor = 3 : i32, fused_activation_function = "RELU", padding = "SAME", stride_h = 4 : i32, stride_w = 5 : i32} : (tensor<1x112x112x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>, tensor<32x3x3x3x!quant.uniform<u8:f32, 0.0078431372549019607:128>>, tensor<32xf32>) -> tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>
  %1 = "tfl.add"(%0, %cst) {fused_activation_function = "NONE"} : (tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>, tensor<32xf32>) -> tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>
  return %1 : tensor<1x56x56x32x!quant.uniform<u8:f32, 0.0078431372549019607:128>>
}
```

**Other info / logs**

@liufengdb fixed a very similar issue in the quantization driver with https://github.com/tensorflow/tensorflow/commit/4a17afaf6e5d891a3e21561fa20ca093fe09b4e2 and https://github.com/tensorflow/tensorflow/commit/2f01bf6606ee6f10c7e57b98d16a3926d2087fd5 so it would be great to integrate the fixes in the `DefaultQuantParamsPass` as well. This pass is very useful for benchmarking quantized models without the need to correctly include fake quant ops in the graph. I tried using the post-training quantization workflow, but in my testing it failed when used together with custom ops.

**Comment by lgeiger** [comment]
@amahendrakar @daverim This has been labeled with `TF 2.3` are there any updates on whether you are still planing to look into a fix for the upcoming release?

**Comment by bhack** [comment]
What is the status of this?

**Comment by lgeiger** [comment]
> What is the status of this?

As far as I can tell, this is still an issue and hasn't been fixed yet.

**Comment by daverim** [comment]
Sorry for the delay on this. We're looking into the issue and trying to
resolve it.

**Comment by daverim** [comment]
Recently tested this with tf-nightly -- seems to be resolved now? The
constants are properly duplicated on the toy model originally posted.

**Comment by lgeiger** [comment]
> Recently tested this with tf-nightly -- seems to be resolved now? The
constants are properly duplicated on the toy model originally posted.

Are you sure? I recently tested the MLIR filechecks again, and it still seems to be an issue or am I missing something?

**Comment by lgeiger** [comment]
> This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you.

I think this is still an issue :)

**Comment by pjpratik** [comment]
Hi @lgeiger 

Could you please test with latest 2.13 version and check if the issue still persists?

Thanks.
